internal/healthchecks/generic/generic.go

```diff
@@ -0,0 +1,198 @@
+package generic
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// AreObjectsHealthy checks if the given resources are healthy.
+// It returns true if all resources are healthy with nil error.
+// It returns false if any of the resources are not healthy, the error contains the GVK + resourceName and the error message of
+// each unhealthy resource.
+//
+// The current list of supported resources is:
+// - Deployments
+// - StatefulSets
+// - DaemonSets
+// - ReplicaSets
+// - Pods
+// - APIServices
+// - CustomResourceDefinitions
+//
+// If the resource is not supported, it is assumed to be healthy.
+func AreObjectsHealthy(ctx context.Context, client client.Client, objects []client.Object) (bool, error) {
+	gvkErrorsMap := make(map[string]string)
```

Comment:
Do we need a map? I wonder if we can just have a `[]error` and use `errors.Join()` when we return it.

And if the order of the `objects` slice is deterministic, then the order of the returned error string will also be deterministic. Is that indirect determinism sufficient?

Reply:
+1 to slice.

Reply:
done, move it to slice